internal/outctx/config.go

```diff
@@ -0,0 +1,131 @@
+	UseSingleKey    bool   `conf:"use_single_key"    validate:"boolean"`
+	AllowMissingKey bool   `conf:"allow_missing_key" validate:"boolean"`
+	SingleKey       string `conf:"time_zone"         validate:"required_if=use_single_key true"`
+	TimeZone        string `conf:"time_zone"         validate:"timezone"`
+}
+
+// Generates configuration struct containing user-defined settings. In addition, sets default values
+// and validates user input.
+//
+// Parameters:
+//   - plugin: Fluent Bit plugin reference
+//
+// Returns:
+//   - S3Config: Configuration based on fluent-bit.conf
+//   - err: All validation errors in config wrapped, parse bool error
+func NewS3Config(plugin unsafe.Pointer) (*S3Config, error) {
+	// Define default values for settings. Setting defaults before validation simplifies validation
+	// configuration, and ensures that default settings are also validated.
+	config := S3Config{
+		// Default Id is uuid to safeguard against s3 filename namespace collision. User may use
+		// multiple collectors to send logs to same s3 path. Id is appended to s3 filename.
+		S3Region:        "us-east-1",
+		S3BucketPrefix:  "logs/",
+		Id:              uuid.New().String(),
+		UseSingleKey:    true,
+		AllowMissingKey: true,
+		SingleKey:       "log",
+		TimeZone:        "America/Toronto",
+	}
+
+	// Map used to loop over user inputs saving a [output.FLBPluginConfigKey] call for each key.
+	// Potential to interate over struct using reflect; however, better to avoid reflect package.
+	pluginSettings := map[string]interface{}{
+		"s3_region":         &config.S3Region,
+		"s3_bucket":         &config.S3Bucket,
+		"s3_bucket_prefix":  &config.S3BucketPrefix,
+		"role_arn":          &config.RoleArn,
+		"id":                &config.Id,
+		"use_single_key":    &config.UseSingleKey,
+		"allow_missing_key": &config.AllowMissingKey,
+		"single_key":        &config.SingleKey,
+		"time_zone":         &config.TimeZone,
+	}
+
+	for settingName, untypedField := range pluginSettings {
+		// [output.FLBPluginConfigKey] retrieves values defined in fluent-bit.conf. Unfortunately,
+		// retrieves all values as strings. If the option is not defined by user, it is set to "".
+		userInput := output.FLBPluginConfigKey(plugin, settingName)
+
+		// If user did not specify a value, do not overwrite default value.
+		if userInput != "" {
+			// Type switch to type parse boolean strings into boolean type. This is neccesary since
+			// all values are provided as strings.
+			switch configField := untypedField.(type) {
+			case *string:
+				*configField = userInput
+			case *bool:
+				// This will throw error if input is "".
+				boolInput, err := strconv.ParseBool(userInput)
+				if err != nil {
+					return nil, fmt.Errorf("error could not parse input %v into bool", userInput)
+				}
+				*configField = boolInput
+			default:
+				return nil, fmt.Errorf("unable to parse type %T", untypedField)
+			}
+		}
```

Comment:
Rather than wrap the almost the entire body, lets reverse the condition and continue early. This should make the code more readable as it reduces indentation and groups logic more tightly (no need to track/remember that `userInput == ""` is the fall through case after reading the if statement body).
i.e.
```
if userInput == "" {
    continue
}

switch configField .......